File: shoring/flywheel/src/main/java/com/panopset/flywheel/Rezop.java
package com.panopset.flywheel;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.StringWriter;
import java.util.List;
import com.panopset.compat.Fileop;
import com.panopset.compat.Streamop;
import com.panopset.compat.Stringop;

public class Rezop {
  
  private Rezop() {}
  
  public static String loadFromRez(Class<?> clazz, String rezPath) {
    return Streamop.getTextFromStream(clazz.getResourceAsStream(rezPath));
  }

  public static List<String> loadListFromTextResource(Class<?> clazz, String rezPath) {
    InputStream is = clazz.getResourceAsStream(rezPath);
    return textStreamToList(is);
  }

  public static List<String> loadListFromTextResource(String rezPath) {
    InputStream is = Rezop.class.getResourceAsStream(rezPath);
    return textStreamToList(is);
  }

  public static List<String> textStreamToList(InputStream is) {
    String lines = Streamop.getTextFromStream(is);
    return Stringop.stringToList(lines);
  }

  public static InputStream getResourceStream(Class<?> clazz, String resourcePath) {
    return clazz.getResourceAsStream(resourcePath);
  }

  public static String getPackageResourcePath(Package pkg) {
    StringWriter sw = new StringWriter();
    sw.append("/");
    sw.append(pkg2path(pkg.getName()));
    return sw.toString();
  }

  public static String pkg2path(String dotName) {
    return dotName.replace(".", "/");
  }

  public static void copyTextResourceToFile(Class<?> clazz, String resourcePath, File file)
      throws IOException {
    Fileop.copyInputStreamToFile(getResourceStream(clazz, resourcePath), file);
  }

  public static void copyTextResourceToFile(Class<?> clazz, String resourcePath, String fileName)
      throws IOException {
    Fileop.copyInputStreamToFile(getResourceStream(clazz, resourcePath), fileName);
  }

  public static String pkg2resourcePath(Class<?> clazz) {
    StringWriter sw = new StringWriter();
    sw.append("/");
    sw.append(clazz.getPackage().getName().replace(".", "/"));
    sw.append("/");
    return sw.toString();
  }
}

File: code/shoring/flywheel/src/main/kotlin/com/panopset/flywheel/Rezop.kt
package com.panopset.flywheel

import com.panopset.compat.Fileop.copyInputStreamToFile
import com.panopset.compat.Streamop.getTextFromStream
import com.panopset.compat.Stringop.stringToList
import java.io.File
import java.io.InputStream
import java.io.StringWriter

object Rezop {
    fun loadFromRez(clazz: Class<*>, rezPath: String): String {
        val inputString = clazz.getResourceAsStream(rezPath) ?: return ""
        return getTextFromStream( inputString)
    }

    fun loadListFromTextResource(clazz: Class<*>, rezPath: String): List<String> {
        val inputStream = clazz.getResourceAsStream(rezPath) ?: return ArrayList()
        return textStreamToList( inputStream)
    }

    fun loadListFromTextResource(rezPath: String): List<String> {
        val inputStream = Rezop::class.java.getResourceAsStream(rezPath) ?: return ArrayList()
        return textStreamToList( inputStream)
    }

    private fun textStreamToList(inputStream: InputStream): List<String> {
        val lines = getTextFromStream( inputStream)
        return stringToList( lines)
    }

    private fun getResourceStream(clazz: Class<*>, resourcePath: String): InputStream? {
        return clazz.getResourceAsStream(resourcePath)
    }

    fun getPackageResourcePath(pkg: Package): String {
        val sw = StringWriter()
        sw.append("/")
        sw.append(pkg2path(pkg.name))
        return sw.toString()
    }

    fun pkg2path(dotName: String): String {
        return dotName.replace(".", "/")
    }

    fun copyTextResourceToFile(clazz: Class<*>, resourcePath: String, file: File) {
        val inputStream = getResourceStream(clazz, resourcePath) ?: return
        copyInputStreamToFile(inputStream, file)
    }

    fun copyTextResourceToFile(clazz: Class<*>, resourcePath: String, fileName: String) {
        val inputStream = getResourceStream(clazz, resourcePath) ?: return
        copyInputStreamToFile(inputStream, fileName)
    }

    fun pkg2resourcePath(clazz: Class<*>): String {
        val sw = StringWriter()
        sw.append("/")
        sw.append(clazz.getPackage().name.replace(".", "/"))
        sw.append("/")
        return sw.toString()
    }
}
